spot-ml/src/main/scala/org/apache/spot/SpotSparkLDAWrapper.scala

```diff
@@ -0,0 +1,154 @@
+package org.apache.spot
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.mllib.clustering.{DistributedLDAModel, EMLDAOptimizer, LDA, OnlineLDAOptimizer}
+import org.apache.spark.mllib.linalg.{Matrix, Vector, Vectors}
+import org.apache.spark.sql.Row
+import org.apache.spark.sql.functions.max
+import java.io.PrintWriter
+import java.io.File
+
+import scala.collection.Map
+import scala.io.Source._
+import scala.sys.process._
+
+/**
+  * Contains routines for LDA including pre and post operations
+  * 1. Creates list of unique documents, words and model based on those two
+  * 2. Processes the model using Spark LDA
+  * 3. Reads Spark LDA results: Topic distributions per document and words per topic
+  * 4. Calculates and returns probability of word given topic: p(w|z)
+  */
+
+object SpotSparkLDAWrapper {
+
+  case class SpotSparkLDAInput(doc: String, word: String, count: Int) extends Serializable
+
+  case class SpotSparkLDAOutput(docToTopicMix: scala.Predef.Map[String, Array[Double]], wordResults: scala.Predef.Map[String, Array[Double]])
+
+
+  def runLDA(docWordCount: RDD[SpotSparkLDAInput],
+             modelFile: String,
+             topicDocumentFile: String,
+             topicWordFile: String,
+             topicCount: Int,
+             localPath: String,
+             ldaPath: String,
+             localUser: String,
+             dataSource: String,
+             ldaSeed: Option[Long],
+             ldaOptimizer: String = "em",
+             ldaAlpha : Double = 2.5,
+             ldaBeta : Double = 1.1,
+             maxIterations: Int = 20 ):   SpotSparkLDAOutput =  {
+
+    // Create word Map Word,Index for further usage
+    val wordDictionary: Map[String, Int] = {
+      val words = docWordCount
+        .cache
+        .map({case SpotSparkLDAInput(doc, word, count) => word})
+        .distinct
+        .collect
+      words.zipWithIndex.toMap
+    }
+
+    val distinctDocument: Array[String] = docWordCount.map({case SpotSparkLDAInput(doc, word, count) => doc}).distinct.collect
+    //distinctDocument.cache()
+
+    // Create document Map Index, Document for further usage
+    val documentDictionary: Map[Int, String] = {
+      distinctDocument
+        //.collect
+        .zipWithIndex
+        .sortBy(_._2)
+        .map(kv => (kv._2, kv._1))
+        .toMap
+    }
+    val docStrToID = documentDictionary.map(_.swap)
+
+    //****Spark LDA implementation****
+
+    //Convert SpotSparkLDAInput into desired format for Spark LDA: (doc, word, count) -> word count per doc, where RDD
+    //is indexed by DocID
+    val wordCountsPerDoc : RDD[(Long, Iterable[(Int, Double)])]
+    = docWordCount.map({case SpotSparkLDAInput(doc, word, count) => (docStrToID(doc).toLong, (wordDictionary(word), count.toDouble))}).groupByKey
+
+    val testDupWords: RDD[(Long, Int)] = docWordCount.map({case SpotSparkLDAInput(doc, word, count) => (docStrToID(doc).toLong, wordDictionary(word))})

```

Comment:
how expensive is this step?  it's a reasonable validation, test step for the inputs,
but if it's not nearly-free, maybe it should go into some kind of unit or regression test to catch bugs like that when testing, but not at every run of the program